server/bitbucket_server/factory.go

```diff
@@ -0,0 +1,14 @@
+package bitbucket_server
+
+import (
+	"fmt"
+
+	bitbucketv1 "github.com/gfleury/go-bitbucket-v1"
+)
+
+func GetBitbucketClient(clientType string, selfHostedURL string, selfHostedAPIURL string, apiClient *bitbucketv1.APIClient) (Client, error) {
+	if clientType == "server" {
+		return newServerClient(selfHostedURL, selfHostedAPIURL, apiClient), nil
+	}
+	return nil, fmt.Errorf("wrong client passed")
+}
```

Comment:
If this package only creates a Bitbucket Server client, and not a Bitbucket Cloud client, what benefit do we get from having this `clientType` parameter?

Reply:
Oh, the idea is to create the Bitbucket Cloud as well! I didn't insert this yet because we plan to do this more consistently on this [issue](https://github.com/mattermost/mattermost-plugin-bitbucket/issues/124).

But if you want, I can add the condition here.

Reply:
I think the function to create the BB Cloud client will be in a separate package though like `bitbucket_cloud`. What do you think? We'll see how the code feels when instantiating either client from the main package

Reply:
We can change the name of this package to just `package bitbucket` and have both on the same package.

so we have:
- `client.go`: generic client with the necessary methods
- `client_server`: concrete server client with the implementation of the Bitbucket Server
- `client_cloud`: concrete cloud client with the implementation of the Bitbucket Cloud

What do you think of this approach?

Reply:
If we think that everything to do with Bitbucket Server can/should fit in its own file, then this works well. But if there are multiple files for each client type, I think we should have separate packages for each client type. What do you think?

So either this that I think you're suggesting

```go
client := bitbucket.NewBBServerClient()
```

or

```go
client := bitbucket_server.NewClient()
```